src/common/buffer.cc

```diff
@@ -1802,13 +1802,33 @@ using namespace ceph;
 
   void buffer::list::claim_prepend(list& bl, unsigned int flags)
   {
-    // steal the other guy's buffers
+    // if the first ptr in buffers has residual data, we must clear it
+    // 1. get the residual ptr of bl
+    const ptr residual_ptr = front();
+
+    if (residual_ptr.offset() != 0) {
+      ptr whole_ptr(residual_ptr.length());
+      whole_ptr.copy_in(0, residual_ptr.length(), residual_ptr.get_raw()->data);
+
+      // 2. erase the incompleted ptr
+      std::list<ptr>::iterator curbuf = _buffers.begin();
+      _len -= (*curbuf).length();
+      _buffers.erase(curbuf);
+      
+      // 3. push front the residual ptr back to bl
+      push_front(whole_ptr);
+    }
+
+    // 4. steal the other guy's buffers
     _len += bl._len;
     if (!(flags & CLAIM_ALLOW_NONSHAREABLE))
       bl.make_shareable();
     _buffers.splice(_buffers.begin(), bl._buffers );
     bl._len = 0;
     bl.last_p = bl.begin();
+
+    // 5. update the last_p point to the lasted ptr
+    last_p = begin();
```

Comment:
i think we just need the step 5 to fix this issue?

and would you please add a test case in `src/test/bufferlist.cc`?